### PR TITLE
FIX Add eagerloaded data to ALL relevant lists

### PR DESCRIPTION
**Pull request by GuySartorelli**
## Description

When a record belongs to multiple many_many (and maybe has_many) relations and you eagerload it and chained relations from it, you can end up in a scenario where the eagerloaded data doesn't get added to some of the relation lists it belongs to.
More details in the linked issue.

## Manual testing steps
1. Add a breakpoint on `DataList::getFinalisedQuery()`
2. Run this code
    ```php
    use SilverStripe\Security\Member;
    $list = Member::get()->eagerLoad('Groups.Permissions');
    foreach ($list as $member) {
        foreach ($member->DirectGroups() as $group) {
            foreach ($group->Permissions() as $perm) {
                echo '';
            }
        }
    }
    ```
3. Check the trace each time the breakpoint is hit. It should only be hit once directly from the above code.

You could also add a breakpoint in the above code itself, and check that each list being looped over (with exception for the top loop) is an instance of `EagerLoadedList`, not `DataList`.

## Issues

- https://github.com/silverstripe/silverstripe-framework/issues/11138

## Pull request checklist

- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [x] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

### CODE REVIEW

File: src/ORM/EagerLoadedList.php [hunk 1]
```diff
@@ -940,7 +940,7 @@ private function setDataObjectEagerLoadedData(int $id, DataObject $item): void
 
     /**
      * Gets the final rows for this list after applying all transformations.
-     * Currently only limit is applied lazily, but others could be done this was as well.
+     * Currently only limit and sort are applied lazily, but filter could be done this was as well in the future.
```
Comment (GuySartorelli) [inline src/ORM/EagerLoadedList.php:943]: This is an unrelated change - noticed that this PHPDoc was lying so I tidied that up while I was here.

**Comment by GuySartorelli** [comment]
Note that I have confirmed the new test fails without this fix.

**Review by emteknetnz — APPROVED** [review]

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Just confirming - did you intentionally target `5`? Seems like this should target `5.1`?

**Comment by GuySartorelli** [comment]
Oops! Good catch. Targetted to `5.1`

**Review by emteknetnz — APPROVED** [review]
